src/runtime.jl

```diff
@@ -137,34 +122,32 @@ function _setup_interpolation(::Type{SyntaxTree}, ex, values)
         end
     end
     if isnothing(graph)
-        graph = ensure_attributes(SyntaxGraph(), kind=Kind, syntax_flags=UInt16, source=SourceAttrType,
-                                  value=Any, name_val=String, scope_layer=LayerId)
+        graph = ensure_attributes(
+            SyntaxGraph(), kind=Kind, syntax_flags=UInt16, source=SourceAttrType,
+            value=Any, name_val=String, scope_layer=LayerId)
     end
     ctx = InterpolationContext(graph, values, Ref(1))
-    return ctx
-end
 
-function _setup_interpolation(::Type{Expr}, ex, values)
-    return ExprInterpolationContext(values, Ref(1))
+    # We must copy the AST into our context to use it as the source reference of
+    # generated expressions.
+    ex1 = copy_ast(ctx, ex)
+    out = _interpolate_ast(ctx, @ast(ctx, ex1, [K"None" ex1]), 0)
+    length(children(out)) === 1 || throw(
+        LoweringError(ex1, "More than one value in bare `\$` expression"))
+    return only(children(out))
 end
 
-function interpolate_ast(::Type{T}, ex, values...) where {T}
-    ctx = _setup_interpolation(T, ex, values)
-
-    # We must copy the AST into our context to use it as the source reference
-    # of generated expressions (and in the Expr case at least, to avoid mutation)
-    ex1 = copy_ast(ctx, ex)
-    if _interp_kind(ex1) == K"$"
-        @assert length(values) == 1
-        vs = values[1]
-        if length(vs) > 1
-            # :($($(xs...))) where xs is more than length 1
-            throw(LoweringError(_to_syntax_tree(ex1),
-                                "More than one value in bare `\$` expression"))
+function interpolate_ast(::Type{Expr}, @nospecialize(ex), values...)
+    ctx = ExprInterpolationContext(values, Ref(1))
+    if ex isa Expr && ex.head === :$
+        @assert length(values) === 1
+        if length(ex.args) !== 1
+            throw(LoweringError(
+                expr_to_syntaxtree(ex), "More than one value in bare `\$` expression"))
```

Comment:
Looks like we need a test to hit this code path now that this error code path is duplicated.

We do have an obscure test case in misc.jl but it's rather hard to find and doesn't include the error message. I'll push an update to that test and also make it work in `expr_compat_mode`